mlir/lib/Gradient/Transforms/ConversionPatterns.cpp

```diff
@@ -220,145 +166,301 @@ struct BackpropOpPattern : public OpConversionPattern<BackpropOp> {
+                unpackMemRef(arg, dataIn[position], callArgs, rewriter, loc, {.zeroOut = true});
+            }
+            index++;
+        }
 
-                SmallVector<int64_t> vectorIndices{3, 0};
+        for (Value qJacobian : op.getQuantumJacobian()) {
+            // Enzyme requires buffers for the primal outputs, but we don't need their values.
+            // We'll need to allocate space for them regardless, so marking them as dupNoNeed will
+            // allow Enzyme to optimize away their computation.
+            auto memrefType = cast<MemRefType>(qJacobian.getType());
+            SmallVector<Value> dynamicDims;
+            for (int64_t dim = 0; dim < memrefType.getRank(); dim++) {
+                if (memrefType.isDynamicDim(dim)) {
+                    Value dimIndex = rewriter.create<index::ConstantOp>(loc, dim);
+                    dynamicDims.push_back(rewriter.create<memref::DimOp>(loc, qJacobian, dimIndex));
+                }
+            }
+            Value result = rewriter.create<memref::AllocOp>(loc, memrefType, dynamicDims);
 
-                Value sizeArray =
-                    rewriter.create<LLVM::ExtractValueOp>(loc, llvmMemrefArg, vectorIndices);
+            unpackMemRef(result, qJacobian, callArgs, rewriter, loc, {.dupNoNeed = true});
+        }
 
-                vectorIndices[0] = 4;
-                Value strideArray =
-                    rewriter.create<LLVM::ExtractValueOp>(loc, llvmMemrefArg, vectorIndices);
+        // The results of backprop are in data in
+        rewriter.create<LLVM::CallOp>(loc, backpropFnDecl, callArgs);
+        rewriter.eraseOp(op);
+        return success();
+    }
 
-                bufferSize = rewriter.create<LLVM::MulOp>(loc, sizeArray, strideArray);
-                bufferSize = rewriter.create<LLVM::AddOp>(loc, offset, bufferSize);
-            }
+  private:
+    static FlatSymbolRefAttr getOrInsertEnzymeGlobal(OpBuilder &builder, ModuleOp moduleOp,
+                                                     const char *globalName)
+    {
+        // Copyright (C) 2023 - Jacob Mai Peng
+        // https://github.com/pengmai/lagrad/blob/main/lib/LAGrad/LowerToLLVM.cpp
+        auto *context = moduleOp.getContext();
+        if (moduleOp.lookupSymbol<LLVM::GlobalOp>(globalName)) {
+            return SymbolRefAttr::get(context, globalName);
+        }
 
-            Value bufferMemSize = rewriter.create<LLVM::MulOp>(loc, bufferSize, memrefArgSize);
+        OpBuilder::InsertionGuard insertGuard(builder);
+        builder.setInsertionPointToStart(moduleOp.getBody());
+        auto shortTy = IntegerType::get(context, 8);
+        builder.create<LLVM::GlobalOp>(moduleOp.getLoc(), shortTy,
+                                       /*isConstant=*/true, LLVM::Linkage::Linkonce, globalName,
+                                       IntegerAttr::get(shortTy, 0));
+        return SymbolRefAttr::get(context, globalName);
+    }
```

Comment:
@josh146 does this copyright notice look good?

Reply:
Looks good to me!